### PR TITLE
AQCU-774: Does not display aging status for DV correctly if day has mixed aging status.

### PR DESCRIPTION
**Pull request by zmoore-usgs**
Issue was that the approval dots were not using the "lowest" approval status on a given day. For example, sometimes a day would be "Working" until 9:00AM and then switch to "In Review" and a point on that day might be rendered as "In Review" but it should be rendered as the lowest approval status on that day which would be "Working".

I am definitely not confident that this is the most efficient way to accomplish this as I'm still learning R so any feedback on that front would be greatly appreciated!

### CODE REVIEW

File: R/utils-json.R [hunk 1]
```diff
@@ -169,47 +169,74 @@ getApprovals <- function(data, chain_nm, legend_nm, appr_var_all, month=NULL, po
...
+
+    review_index<- apply(review_dates, 1, function(d, points){
+      which(points$time >= d[1] & points$time <= d[2])}, 
+      points=points_no_times)
+
+    approved_index<- apply(approved_dates, 1, function(d, points){
+      which(points$time >= d[1] & points$time <= d[2])}, 
+      points=points_no_times)
+
+    if(is.list(working_index)){
```
Comment (lindsayplatt) [inline R/utils-json.R]: I would recommend not using `is.list()`. It's tricky -- data.frames are technically "lists" under the hood. Try `class(working_index) == "list")` instead.

Trickiness:

``` r
ex <- data.frame(a=1,b=2)
is.list(ex)
[1] TRUE
```

File: R/utils-json.R [hunk 2]
```diff
@@ -169,47 +169,74 @@ getApprovals <- function(data, chain_nm, legend_nm, appr_var_all, month=NULL, po
...
+    if(is.list(review_index)){
+      review_index <- unique(unlist(review_index, recursive=FALSE))
+    }
+
+    if(is.list(approved_index)){
+      approved_index <- unique(unlist(approved_index, recursive=FALSE))
+    }
+    
+    review_index <- setdiff(review_index, working_index)
+    approved_index <- setdiff(approved_index, review_index)
```
Comment (lindsayplatt) [inline R/utils-json.R:198]: I've never seen that function - handy!!

File: R/utils-json.R [hunk 3]
```diff
@@ -169,47 +169,74 @@ getApprovals <- function(data, chain_nm, legend_nm, appr_var_all, month=NULL, po
...
-        points=points)
-      
-      if(is.list(date_index)){
-        date_index_list <- date_index
-      } else {
-        date_index_list <- list(date_index)
-      }
+    approved_dates <- getApprovalDates(data, chain_nm, "Approved")
+    review_dates <-getApprovalDates(data, chain_nm, "In Review")
+    working_dates <- getApprovalDates(data, chain_nm, "Working")
```
Comment (lindsayplatt) [inline R/utils-json.R]: :+1: for removing that top for loop

File: R/utils-json.R [hunk 4]
```diff
@@ -169,47 +169,74 @@ getApprovals <- function(data, chain_nm, legend_nm, appr_var_all, month=NULL, po
...
       
+    if(subsetByMonth){
+      points <- subsetByMonth(getTimeSeries(data, chain_nm), month)
+      points_no_times <- points
```
Comment (lindsayplatt) [inline R/utils-json.R]: 178 & 179 can be moved outside of the logic because they are the same as 182 & 183

**Comment by lindsayplatt** [comment]
It seems like a lot of the code finding the approvals (where you do the same thing for inreview, working, and approved) can be pulled out into one function where the argument is the approval type. That way we can get rid of some duplicated code.

File: R/utils-json.R [hunk 5]
```diff
@@ -169,47 +169,74 @@ getApprovals <- function(data, chain_nm, legend_nm, appr_var_all, month=NULL, po
...
+      which(points$time >= d[1] & points$time <= d[2])}, 
+      points=points_no_times)
+
+    review_index<- apply(review_dates, 1, function(d, points){
+      which(points$time >= d[1] & points$time <= d[2])}, 
+      points=points_no_times)
+
+    approved_index<- apply(approved_dates, 1, function(d, points){
+      which(points$time >= d[1] & points$time <= d[2])}, 
+      points=points_no_times)
```
Comment (lindsayplatt) [inline R/utils-json.R]: these apply statements can be in a single function

File: R/utils-json.R (hunk 1 above)
Comment (lindsayplatt) [reply R/utils-json.R]: these if statements can go in the same function as the applys starting on line 186

**Comment by zmoore-usgs** [comment]
Whoops, used the wrong ticket number in my last commit message.  Should be AQCU-774.

File: R/utils-json.R [hunk 6]
```diff
@@ -164,52 +164,66 @@ getEstimatedDates <- function(data, chain_nm, time_data){
   return(date_index)
 }
 
+getApprovalIndex <- function(data, points, chain_nm, approval, subsetByMonth=FALSE) {
```
Comment (lindsayplatt) [inline R/utils-json.R:167]: :+1:

**Pull request merged**